File: backend/src/layered_span_studio_backend/api/labels.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query, status

from layered_span_studio_backend.core.dependencies import get_current_user, get_settings
from layered_span_studio_backend.models.labels import (
    LabelExamplesResponse,
    LabelExamplesSampleMode,
    LabelSurfaceGroupsResponse,
    LabelExamplesStatusFilter,
    LabelListResponse,
    LabelSyncIn,
)
from layered_span_studio_backend.services import labels_service

router = APIRouter(
    prefix="/projects/{project_id}/labels",
    tags=["labels"],
    dependencies=[Depends(get_current_user)],
)


@router.get("", response_model=LabelListResponse)
def list_labels(project_id: str, settings=Depends(get_settings)):
    try:
        response = labels_service.list_labels_state(settings, project_id)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))
    return response


@router.put("", response_model=LabelListResponse)
def save_labels(project_id: str, payload: LabelSyncIn, settings=Depends(get_settings)):
    try:
        response = labels_service.save_labels(
            settings,
            project_id,
            [label.model_dump(mode="json") for label in payload.labels],
            payload.base_revision,
        )
    except ValueError as exc:
        message = str(exc)
        status_code = status.HTTP_400_BAD_REQUEST
        if "Project not found" in message or "Label not found" in message:
            status_code = status.HTTP_404_NOT_FOUND
        if message == "Label revision mismatch":
            status_code = status.HTTP_409_CONFLICT
        raise HTTPException(status_code=status_code, detail=message)
    return response


@router.get("/{label_id}/examples", response_model=LabelExamplesResponse)
def list_label_examples(
    project_id: str,
    label_id: str,
    offset: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    status_filter: LabelExamplesStatusFilter = Query(LabelExamplesStatusFilter.verified, alias="status"),
    sample: LabelExamplesSampleMode = Query(LabelExamplesSampleMode.sequential),
    seed: int | None = Query(None),
    context_window: int = Query(20, ge=0, le=200),
    settings=Depends(get_settings),
):
    try:
        return labels_service.list_label_examples(
            settings,
            project_id,
            label_id,
            offset,
            limit,
            status_filter.value,
            sample.value,
            seed,
            context_window,
        )
    except ValueError as exc:
        message = str(exc)
        status_code = status.HTTP_400_BAD_REQUEST
        if message in {"Project not found", "Label not found"}:
            status_code = status.HTTP_404_NOT_FOUND
        raise HTTPException(status_code=status_code, detail=message)


@router.get("/{label_id}/surface-groups", response_model=LabelSurfaceGroupsResponse)
def list_label_surface_groups(
    project_id: str,
    label_id: str,
    offset: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
    status_filter: LabelExamplesStatusFilter = Query(LabelExamplesStatusFilter.verified, alias="status"),
    context_window: int = Query(20, ge=0, le=200),
    exclude_annotation_id: str | None = Query(None),
    settings=Depends(get_settings),
):
    try:
        return labels_service.list_label_surface_groups(
            settings,
            project_id,
            label_id,
            offset,
            limit,
            status_filter.value,
            context_window,
            exclude_annotation_id,
        )
    except ValueError as exc:
        message = str(exc)
        status_code = status.HTTP_400_BAD_REQUEST
        if message in {"Project not found", "Label not found"}:
            status_code = status.HTTP_404_NOT_FOUND
        raise HTTPException(status_code=status_code, detail=message)

File: frontend/src/api/labels.ts
import type {
  LabelListResponse,
  LabelSurfaceGroupsResponse,
  SaveProjectLabelInput,
} from "../api-contract";
import { toJsonObject } from "../utils";
import { ApiError, client, unwrapData } from "./client";

function assertLabelListResponseRevision(payload: LabelListResponse): LabelListResponse {
  if (!payload.revision || payload.revision.trim().length === 0) {
    throw new ApiError("Labels revision が取得できなかった", 500);
  }
  return payload;
}

export function listLabels(projectId: string) {
  return unwrapData<LabelListResponse>(client.GET("/projects/{project_id}/labels", {
    params: { path: { project_id: projectId } },
  })).then(assertLabelListResponseRevision);
}

export function saveProjectLabels(
  projectId: string,
  labels: Array<Pick<SaveProjectLabelInput, "name" | "color" | "description" | "shortcut" | "meta"> & { id: string | null }>,
  baseRevision: string,
) {
  return unwrapData<LabelListResponse>(client.PUT("/projects/{project_id}/labels", {
    params: { path: { project_id: projectId } },
    body: {
      base_revision: baseRevision,
      labels: labels.map((label) => ({
        ...label,
        meta: toJsonObject(label.meta ?? null),
      })),
    },
  })).then(assertLabelListResponseRevision);
}

type LabelSurfaceGroupOptions = {
  offset?: number;
  limit?: number;
  status?: string;
  contextWindow?: number;
  excludeAnnotationId?: string | null;
};

export function listLabelSurfaceGroups(
  projectId: string,
  labelId: string,
  options?: LabelSurfaceGroupOptions,
) {
  return unwrapData<LabelSurfaceGroupsResponse>(client.GET("/projects/{project_id}/labels/{label_id}/surface-groups", {
    params: {
      path: { project_id: projectId, label_id: labelId },
      query: {
        offset: options?.offset ?? 0,
        limit: options?.limit ?? 50,
        status: (options?.status ?? "verified") as "pending" | "verified" | "all",
        context_window: options?.contextWindow ?? 20,
        exclude_annotation_id: options?.excludeAnnotationId ?? undefined,
      },
    },
  }));
}
